rlmolecule/gym/gym_state.py

```diff
@@ -0,0 +1,56 @@
+import base64
+from copy import deepcopy
+import logging
+import pickle
+from typing import Sequence
+
+
+import numpy as np
+import gym
+
+from rlmolecule.tree_search.graph_search_state import GraphSearchState
+from rlmolecule.gym.alphazero_gym import AlphaZeroGymEnv
+
+logger = logging.getLogger(__name__)
+
+
+class GymEnvState(GraphSearchState):
+    """Gyn env state implementation that maps the gym API to the GraphSearchState
+    interface.  Should work generically for any gym env."""
+
+    def __init__(self, 
+                 env: AlphaZeroGymEnv,
+                 step_count: int,
+                 step_reward: float,
+                 cumulative_reward: float,
+                 done: bool,
+                 meta: dict = {}) -> None:
+        assert isinstance(env.action_space, gym.spaces.Discrete)
+        self.env = deepcopy(env)
+        self.step_count = step_count
+        self.step_reward = step_reward
+        self.cumulative_reward = cumulative_reward
+        self.done = done
+        self.meta = meta
+
+    def __repr__(self) -> str:
+        return (self.env.get_obs(), self.step_count).__repr__()
+
+    def equals(self, other: any) -> bool:
+        are_close = np.all(np.isclose(self.env.get_obs(), other.env.get_obs()))
+        same_time = (self.step_count == other.step_count)
+        return are_close and same_time
+
+    def hash(self) -> int:
+        return hash(self.__repr__())
```

Comment:
`__repr__` of a numpy array probably looses a lot of information that you might need -- it only prints the first & last few rows and columns.
I'm assuming Charles wrote [this hash](https://github.com/NREL/rlmolecule/blob/1732e57acc1e0c990ad79b875524dd30573e2dea/rlmolecule/molecule/molecule_state.py#L54)

I did some quick [googling](https://madhatter106.github.io/DataScienceCorner/posts/speeding-up-numpy-array-comparison/), this might work?
`hash(self.env.get_obs().to_bytes()) ^ (13 * self.step_count)`

Reply:
ok, yeah i'm with you on the np array repr, hashing the bytestring makes sense.  but what is this bitwise xor stuff doing?

Reply:
no idea, i actually mis-read that as an exponentiation 😦 . @CharlesEdisonTripp, did you add that to the molecule state?
We just need a way to add that info into the hash. Maybe `hash((self.env.get_obs().to_bytes(), self.step_count))`?

Reply:
yeah, i think that should work!  charles i am also curious about your xor voodoo : )

Reply:
Yeah was just an easy way to add in the _forced_terminal value to the hash of the string repr. A simpler and probably just as effective way would have been to add 13 * _forced_terminal (because _forced_terminal is boolean python will treat it as a 1 or 0 here). The xor is mostly just fancy in this case.. although I wanted to avoid the possibility of overflowing to beyond 64 bits.

A more general way to combine hash functions that usually works pretty well is to add them, multiplying each hash by some prime number, for example:  hash(a) + 13 * hash(b) + 31 * hash(c). In python's case of big integers, you might round it out with a bitwise and to clip it down to 64 bits: (hash(a) + 13 * hash(b) + 31 * hash(c)) & (2 ** 64 - 1).

Reply:
You could also avoid the clipping problem with XOR's, likely while retaining good hash properties:
hash(a) ^ (13 * hash(b)) ^ (31 * hash(c)). Extra points if you use a big "good hash table" prime (for example, see https://planetmath.org/goodhashtableprimes):

hash(a) ^ (805306457 * hash(b)) ^ (402653189 * hash(c))

If the hash functions are really orthogonal, you could safely xor them together without the primes:

hash(a) ^ hash(b) ^ hash(c)

But the issue with doing this typically is that if these were (say) all strings with the same hash function, simply xoring them together will give the same hash value for different permutations of the three strings. That's why we multiply by a prime. We use a prime because it typically gives better dispersion over the table than an ordinary number, particularly with power-of-two sized hash tables, which are typical for performance, memory management, and implementation reasons.

Reply:
Just to follow up, python does some binary XOR under the hood when you hash a tuple